waltz-tools/src/main/java/com/wepay/waltz/tools/cluster/ClusterCli.java

```diff
@@ -364,6 +375,52 @@ private void buildStoragePartitionValidationResult(int numPartitions,
             }
         }
 
+        private void buildStorageRecoveryCompleteValidationResult(int numPartitions,
+                                                                  ZNode partitionRoot,
+                                                                  ZooKeeperClient zkClient,
+                                                                  List<PartitionValidationResults> partitionsValidationResultsList) {
+            for (int partitionId = 0; partitionId < numPartitions; partitionId++) {
+                String error = "";
+                ZNode zNode = new ZNode(partitionRoot, Integer.toString(partitionId));
+                Map<ReplicaId, ReplicaState> replicaState;
+                Long sessionId;
+
+                try {
+                    PartitionMetadata partitionMetadata = zkClient.getData(zNode, PartitionMetadataSerializer.INSTANCE).value;
+                    replicaState = partitionMetadata.replicaStates;
+                    sessionId = partitionMetadata.sessionId;
+                } catch (KeeperException | ZooKeeperClientException e) {
+                    replicaState = Collections.emptyMap();
+                    sessionId = null;
+                    error = "Cannot obtain replica states";
+                }
+
+                for (Map.Entry<ReplicaId, ReplicaState> entry : replicaState.entrySet()) {
+                    boolean recoveryComplete =
```

Comment:
In the recovery process, the sessionId is first updated in partitionMetadata at the beginning, and each replica session's sessionId and closing HWM won't be updated until the recovery finished. So here I check for these two conditions for the complete.

Reply:
Does it cover the case where recovery didn't start at all?

Reply:
> Does it cover the case where recovery didn't start at all?

@abhipardhu No, I don't think so.. I think the verification in this PR only checks for if the replicas are in `recovery completed` or `recovery ongoing` status.
To check if certain recovery was triggered/completed, the verification needs to know the sessionId before the recovery to compare with, but we didn't take that parameter here. 
Do you think we need to add such option here? Like add a `-s` parameter for sessionId, and verify if all replicas have an updated sessionId compared to that input?

Reply:
@abhipardhu : Are you referring to the case where replica assignment of a storage node is updated with a partition_Id, but that replica is not part of the corresponding partition's replica state?

For eg: In the below scenario, the replica states doesn't show `storage04:300`.

store [/waltz/store] replica and group assignments:
  storage01:300 = [0], GroupId: 1
  storage02:300 = [0] GroupId: 2
  storage03:300 = [0], GroupId: 0
  storage04:300 = [0], GroupId: 0
store [/waltz/store] connections:
  storage01:300 has admin port: 301
  storage02:300 has admin port: 301
  storage03:300 has admin port: 301
store [/waltz/store/partition/0] replica states:
  ReplicaId(0,storage01:300), SessionId: 2, closingHighWaterMark: UNRESOLVED
  ReplicaId(0,storage02:300), SessionId: 2, closingHighWaterMark: UNRESOLVED
  ReplicaId(0,storage03:300), SessionId: 2, closingHighWaterMark: UNRESOLVED

Reply:
@abhipardhu @smithakoduri Add verification for consistency between `replicaAssignments` vs `replicaStates`, plz take a 👀